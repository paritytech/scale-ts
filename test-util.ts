/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
