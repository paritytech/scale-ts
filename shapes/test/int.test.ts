import * as $ from "../../mod.ts"
import { testInvalid, testShape } from "../../test-util.ts"

function generateIntTests(signed: boolean, size: number) {
  const s = BigInt(size)
  return signed
    ? [
      -1n * 2n ** (s - 1n), // min value
      -1n * 2n ** (s / 2n - 1n), // min value of last size
      -1n,
      0n,
      1n,
      2n ** (s / 2n - 1n) - 1n, // max value of last size
      2n ** (s - 1n) - 1n, // max value
    ]
    : [
      0n,
      1n,
      2n ** (s / 2n) - 1n, // max value of last size
      2n ** s - 1n, // max value
    ]
}

testShape($.u8, generateIntTests(false, 8).map(Number))
testShape($.u16, generateIntTests(false, 16).map(Number))
testShape($.u32, generateIntTests(false, 32).map(Number))
testShape($.u64, generateIntTests(false, 64).map(BigInt))
testShape($.u128, generateIntTests(false, 128).map(BigInt))
testShape($.u256, generateIntTests(false, 256).map(BigInt))

testShape($.i8, generateIntTests(true, 8).map(Number))
testShape($.i16, generateIntTests(true, 16).map(Number))
testShape($.i32, generateIntTests(true, 32).map(Number))
testShape($.i64, generateIntTests(true, 64).map(BigInt))
testShape($.i128, generateIntTests(true, 128).map(BigInt))
testShape($.i128, generateIntTests(true, 128).map(BigInt))
testShape($.i256, generateIntTests(true, 256).map(BigInt))

function generateIntInvalids(signed: boolean, size: number, fn: typeof Number | typeof BigInt) {
  const s = BigInt(size)
  return [
    null,
    undefined,
    {},
    "abc",
    NaN,
    1.2,
    fn === BigInt ? 0 : 0n,
    signed ? fn(-1n * 2n ** (s - 1n) - 1n) : fn(-1n),
    signed ? fn(2n ** (s - 1n)) : fn(2n ** s),
  ]
}

testInvalid($.u8, generateIntInvalids(false, 8, Number))
testInvalid($.u16, generateIntInvalids(false, 16, Number))
testInvalid($.u32, generateIntInvalids(false, 32, Number))
testInvalid($.u64, generateIntInvalids(false, 64, BigInt))
testInvalid($.u128, generateIntInvalids(false, 128, BigInt))
testInvalid($.u256, generateIntInvalids(false, 256, BigInt))

testInvalid($.i8, generateIntInvalids(true, 8, Number))
testInvalid($.i16, generateIntInvalids(true, 16, Number))
testInvalid($.i32, generateIntInvalids(true, 32, Number))
testInvalid($.i64, generateIntInvalids(true, 64, BigInt))
testInvalid($.i128, generateIntInvalids(true, 128, BigInt))
testInvalid($.i128, generateIntInvalids(true, 128, BigInt))
testInvalid($.i256, generateIntInvalids(true, 256, BigInt))
