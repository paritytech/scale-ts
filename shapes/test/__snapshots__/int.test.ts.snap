export const snapshot = {};

snapshot[`\$.u8 0 1`] = `00`;

snapshot[`\$.u8 1 1`] = `01`;

snapshot[`\$.u8 15 1`] = `0f`;

snapshot[`\$.u8 255 1`] = `ff`;

snapshot[`\$.u16 0 1`] = `
00
00
`;

snapshot[`\$.u16 1 1`] = `
01
00
`;

snapshot[`\$.u16 255 1`] = `
ff
00
`;

snapshot[`\$.u16 65535 1`] = `
ff
ff
`;

snapshot[`\$.u32 0 1`] = `
00
00
00
00
`;

snapshot[`\$.u32 1 1`] = `
01
00
00
00
`;

snapshot[`\$.u32 65535 1`] = `
ff
ff
00
00
`;

snapshot[`\$.u32 4294967295 1`] = `
ff
ff
ff
ff
`;

snapshot[`\$.u64 0n 1`] = `
00
00
00
00
00
00
00
00
`;

snapshot[`\$.u64 1n 1`] = `
01
00
00
00
00
00
00
00
`;

snapshot[`\$.u64 4294967295n 1`] = `
ff
ff
ff
ff
00
00
00
00
`;

snapshot[`\$.u64 18446744073709551615n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.u128 0n 1`] = `
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.u128 1n 1`] = `
01
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.u128 18446744073709551615n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
00
00
00
00
00
00
00
00
`;

snapshot[`\$.u128 340282366920938463463374607431768211455n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.u256 0n 1`] = `
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.u256 1n 1`] = `
01
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.u256 340282366920938463463374607431768211455n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.u256 115792089237316195423570985008687907853269984665640564039457584007913129639935n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.i8 -128 1`] = `80`;

snapshot[`\$.i8 -8 1`] = `f8`;

snapshot[`\$.i8 -1 1`] = `ff`;

snapshot[`\$.i8 0 1`] = `00`;

snapshot[`\$.i8 1 1`] = `01`;

snapshot[`\$.i8 7 1`] = `07`;

snapshot[`\$.i8 127 1`] = `7f`;

snapshot[`\$.i16 -32768 1`] = `
00
80
`;

snapshot[`\$.i16 -128 1`] = `
80
ff
`;

snapshot[`\$.i16 -1 1`] = `
ff
ff
`;

snapshot[`\$.i16 0 1`] = `
00
00
`;

snapshot[`\$.i16 1 1`] = `
01
00
`;

snapshot[`\$.i16 127 1`] = `
7f
00
`;

snapshot[`\$.i16 32767 1`] = `
ff
7f
`;

snapshot[`\$.i32 -2147483648 1`] = `
00
00
00
80
`;

snapshot[`\$.i32 -32768 1`] = `
00
80
ff
ff
`;

snapshot[`\$.i32 -1 1`] = `
ff
ff
ff
ff
`;

snapshot[`\$.i32 0 1`] = `
00
00
00
00
`;

snapshot[`\$.i32 1 1`] = `
01
00
00
00
`;

snapshot[`\$.i32 32767 1`] = `
ff
7f
00
00
`;

snapshot[`\$.i32 2147483647 1`] = `
ff
ff
ff
7f
`;

snapshot[`\$.i64 -9223372036854775808n 1`] = `
00
00
00
00
00
00
00
80
`;

snapshot[`\$.i64 -2147483648n 1`] = `
00
00
00
80
ff
ff
ff
ff
`;

snapshot[`\$.i64 -1n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.i64 0n 1`] = `
00
00
00
00
00
00
00
00
`;

snapshot[`\$.i64 1n 1`] = `
01
00
00
00
00
00
00
00
`;

snapshot[`\$.i64 2147483647n 1`] = `
ff
ff
ff
7f
00
00
00
00
`;

snapshot[`\$.i64 9223372036854775807n 1`] = `
ff
ff
ff
ff
ff
ff
ff
7f
`;

snapshot[`\$.i128 -170141183460469231731687303715884105728n 1`] = `
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
80
`;

snapshot[`\$.i128 -9223372036854775808n 1`] = `
00
00
00
00
00
00
00
80
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.i128 -1n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.i128 0n 1`] = `
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.i128 1n 1`] = `
01
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.i128 9223372036854775807n 1`] = `
ff
ff
ff
ff
ff
ff
ff
7f
00
00
00
00
00
00
00
00
`;

snapshot[`\$.i128 170141183460469231731687303715884105727n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
7f
`;

snapshot[`\$.i128 -170141183460469231731687303715884105728n 2`] = `
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
80
`;

snapshot[`\$.i128 -9223372036854775808n 2`] = `
00
00
00
00
00
00
00
80
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.i128 -1n 2`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.i128 0n 2`] = `
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.i128 1n 2`] = `
01
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.i128 9223372036854775807n 2`] = `
ff
ff
ff
ff
ff
ff
ff
7f
00
00
00
00
00
00
00
00
`;

snapshot[`\$.i128 170141183460469231731687303715884105727n 2`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
7f
`;

snapshot[`\$.i256 -57896044618658097711785492504343953926634992332820282019728792003956564819968n 1`] = `
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
80
`;

snapshot[`\$.i256 -170141183460469231731687303715884105728n 1`] = `
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
80
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.i256 -1n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.i256 0n 1`] = `
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.i256 1n 1`] = `
01
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.i256 170141183460469231731687303715884105727n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
7f
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`\$.i256 57896044618658097711785492504343953926634992332820282019728792003956564819967n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
7f
`;

snapshot[`\$.u8 invalid null 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u8 invalid undefined 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u8 invalid {} 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u8 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u8 invalid NaN 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.u8 invalid 1.2 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.u8 invalid 0n 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u8 invalid -1 1`] = `ShapeAssertError: value < 0`;

snapshot[`\$.u8 invalid 256 1`] = `ShapeAssertError: value > 255`;

snapshot[`\$.u16 invalid null 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u16 invalid undefined 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u16 invalid {} 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u16 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u16 invalid NaN 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.u16 invalid 1.2 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.u16 invalid 0n 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u16 invalid -1 1`] = `ShapeAssertError: value < 0`;

snapshot[`\$.u16 invalid 65536 1`] = `ShapeAssertError: value > 65535`;

snapshot[`\$.u32 invalid null 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u32 invalid undefined 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u32 invalid {} 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u32 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u32 invalid NaN 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.u32 invalid 1.2 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.u32 invalid 0n 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.u32 invalid -1 1`] = `ShapeAssertError: value < 0`;

snapshot[`\$.u32 invalid 4294967296 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.u64 invalid null 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u64 invalid undefined 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u64 invalid {} 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u64 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u64 invalid NaN 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u64 invalid 1.2 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u64 invalid 0 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u64 invalid -1n 1`] = `ShapeAssertError: value < 0n`;

snapshot[`\$.u64 invalid 18446744073709551616n 1`] = `ShapeAssertError: value > 18446744073709551615n`;

snapshot[`\$.u128 invalid null 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u128 invalid undefined 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u128 invalid {} 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u128 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u128 invalid NaN 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u128 invalid 1.2 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u128 invalid 0 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u128 invalid -1n 1`] = `ShapeAssertError: value < 0n`;

snapshot[`\$.u128 invalid 340282366920938463463374607431768211456n 1`] = `ShapeAssertError: value > 340282366920938463463374607431768211455n`;

snapshot[`\$.u256 invalid null 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u256 invalid undefined 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u256 invalid {} 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u256 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u256 invalid NaN 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u256 invalid 1.2 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u256 invalid 0 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.u256 invalid -1n 1`] = `ShapeAssertError: value < 0n`;

snapshot[`\$.u256 invalid 115792089237316195423570985008687907853269984665640564039457584007913129639936n 1`] = `ShapeAssertError: value > 115792089237316195423570985008687907853269984665640564039457584007913129639935n`;

snapshot[`\$.i8 invalid null 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i8 invalid undefined 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i8 invalid {} 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i8 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i8 invalid NaN 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.i8 invalid 1.2 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.i8 invalid 0n 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i8 invalid -129 1`] = `ShapeAssertError: value < -128`;

snapshot[`\$.i8 invalid 128 1`] = `ShapeAssertError: value > 127`;

snapshot[`\$.i16 invalid null 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i16 invalid undefined 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i16 invalid {} 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i16 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i16 invalid NaN 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.i16 invalid 1.2 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.i16 invalid 0n 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i16 invalid -32769 1`] = `ShapeAssertError: value < -32768`;

snapshot[`\$.i16 invalid 32768 1`] = `ShapeAssertError: value > 32767`;

snapshot[`\$.i32 invalid null 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i32 invalid undefined 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i32 invalid {} 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i32 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i32 invalid NaN 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.i32 invalid 1.2 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.i32 invalid 0n 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.i32 invalid -2147483649 1`] = `ShapeAssertError: value: invalid int`;

snapshot[`\$.i32 invalid 2147483648 1`] = `ShapeAssertError: value > 2147483647`;

snapshot[`\$.i64 invalid null 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i64 invalid undefined 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i64 invalid {} 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i64 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i64 invalid NaN 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i64 invalid 1.2 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i64 invalid 0 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i64 invalid -9223372036854775809n 1`] = `ShapeAssertError: value < -9223372036854775808n`;

snapshot[`\$.i64 invalid 9223372036854775808n 1`] = `ShapeAssertError: value > 9223372036854775807n`;

snapshot[`\$.i128 invalid null 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid undefined 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid {} 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid NaN 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid 1.2 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid 0 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid -170141183460469231731687303715884105729n 1`] = `ShapeAssertError: value < -170141183460469231731687303715884105728n`;

snapshot[`\$.i128 invalid 170141183460469231731687303715884105728n 1`] = `ShapeAssertError: value > 170141183460469231731687303715884105727n`;

snapshot[`\$.i128 invalid null 2`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid undefined 2`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid {} 2`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid "abc" 2`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid NaN 2`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid 1.2 2`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid 0 2`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i128 invalid -170141183460469231731687303715884105729n 2`] = `ShapeAssertError: value < -170141183460469231731687303715884105728n`;

snapshot[`\$.i128 invalid 170141183460469231731687303715884105728n 2`] = `ShapeAssertError: value > 170141183460469231731687303715884105727n`;

snapshot[`\$.i256 invalid null 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i256 invalid undefined 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i256 invalid {} 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i256 invalid "abc" 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i256 invalid NaN 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i256 invalid 1.2 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i256 invalid 0 1`] = `ShapeAssertError: typeof value !== "bigint"`;

snapshot[`\$.i256 invalid -57896044618658097711785492504343953926634992332820282019728792003956564819969n 1`] = `ShapeAssertError: value < -57896044618658097711785492504343953926634992332820282019728792003956564819968n`;

snapshot[`\$.i256 invalid 57896044618658097711785492504343953926634992332820282019728792003956564819968n 1`] = `ShapeAssertError: value > 57896044618658097711785492504343953926634992332820282019728792003956564819967n`;
